### ISSUE TITLE
Hook into Windows Explorer

### ISSUE THREAD

**Issue by denosawr** [TranslucentTB/TranslucentTB#134]
As opposed to redrawing the taskbar every `x` milliseconds, we should hook into Windows Explorer and potentially only redraw the taskbar when WinExp calls SWCA.

The `easyhook` package allows for easy DLL injection (without administrator permission) and can hook a function to be called whenever Explorer calls `SetWindowCompositionAttribute` (aka redrawing the taskbar).

**Comment by denosawr** [comment]
There are two current ways to implement this:
* We can use Named Pipes to communicate between Explorer and TranslucentTB. Whenever Explorer calls SWCA, Explorer can communicate to TTB which will appropriately redraw the taskbar
* We can make Explorer redraw the taskbar. It will load the configuration from a memory-mapped file which is written and updated by the main TTB application.

**Comment by denosawr** [comment]
Also, as this will involve a rather large rewrite I will create a new feature branch `feature-explorerinjection` which can be PR'd into `store` when finished.

**Comment by denosawr** [comment]
@ethanhs @sylveon opinions?

**Comment by sylveon** [comment]
I'd rather get a stable update for the Windows Store before we initiate any kind of rewrite

**Comment by denosawr** [comment]
So V5?

**Comment by denosawr** [comment]
I personally would prefer to have it done for V4, considering there are not too many milestones to complete for V4 left.

**Comment by ethanhs** [comment]
I think V4 is getting pretty stable, and I'd rather not rock the boat.

**Comment by denosawr** [comment]
Then I'll set it to V5.

**Comment by denosawr** [comment]
As for memory-mapped files, we have the problem of them being accessed simultaneously by the client and hook... in the end, it may be easier to keep an internal object storing data and use IPC (NamedPipes) to update that data, from another thread.

**Comment by L0laapk3** [comment]
I assume this will solve the flickering of the taskbar to the default color when opening or closing the start menu and that I don't need to create a new issue? its rather distracting :s

**Comment by sylveon** [comment]
It will, but this (as mentionned before) won't make it to V4.

**Comment by ethanhs** [comment]
We should add the update frequency as a config option for now.

**Comment by denosawr** [comment]
@ethanhs Implemented option for config file refresh rate changing here: https://github.com/TranslucentTB/TranslucentTB/commit/8185041c98045cec184a30e93fb7e7bea3f174d9

**Comment by sylveon** [comment]
Instead of using EasyHook, I'd rather use Detours, which is a Microsoft library for that kind of thing licensed under MIT.

By using Detours, we get to be able to live along other hook programs also using Detours, instead of fighting them (hooks into explorer are common, so this is nice).

It's also a DLL injection library made by the writers of the operating system, so way less possibility of having bugs and potentially updated faster if an update to the operating system breaks it (since Detours is used a lot internally)

**Comment by denosawr** [comment]
I'm not aware of EasyHook, but I know for sure that Detours is not available for Windows Store apps.

**Comment by denosawr** [comment]
Also, I think it's worth stating that EasyHook (which is also MIT) is (also?) a fairly well maintained project. Could you please elaborate on how we would have to 'fight hooks'? All we'd be doing is injecting a library which would overwrite SetWindowCompositionAttribute, I'd assume.

**Comment by sylveon** [comment]
Windows explorer is a pretty popular thing to hook, and quite a few other tools (classic shell, startisback, etc...) also inject and overwrite SetWindowCompositionAtteibute

**Comment by denosawr** [comment]
If that is the case (Desktop Bridge allows for the full Win32 API to be used), I'm completely happy using Detours.

(Furthermore, in direct messages Sylveon also mentioned that Detours supports ARM, which is a fairly nice thing to have for futureproofing).

**Comment by Blaze5565** [comment]
I'm not particularly knowledgeable here, but as far as I can tell, that would mean that those with animated wallpapers (such as with Wallpaper Engine) would have a static taskbar area that could look quite off. Perhaps an option to keep the current setup would be a good idea, if there isn't another option?

**Comment by Blaze5565** [comment]
Well, from what I can tell, this planned update would mean that the taskbar is only redrawn when SWCA is called. I'm not sure, but wouldn't that mean that an animated desktop wallpaper playing behind the taskbar wouldn't be animated in the taskbar region, and would only play/update when the taskbar needs to be redrawn?

**Comment by denosawr** [comment]
So, what actually happens here isn't that we manually draw the background of the taskbar - we actually let Windows determine the exact pixels you see on the taskbar. Instead, we set the "Window Composition Attribute" of the taskbar. Normally, the taskbar has a set WCA, and from there Windows will determine what pixels to use following the settings set by the WCA. When we say "redraw the taskbar", we actually mean to set a custom WCA (as Windows likes to set its own WCA quite often, and therefore we have to overwrite that) so that Windows can draw the taskbar how you like it.

TL;DR: We don't draw the pixels behind the taskbar, but instead let Windows do it. So, the proposed options here won't create for 'lagginess' with animated videos.

**Comment by sylveon** [comment]
I've got a working implementation in the `develop` branch. If you're up to the task and want to try it, you can build it yourself (using the README from that branch), or just use [the binaries I built](https://github.com/TranslucentTB/TranslucentTB/files/2429603/ttb-hook.zip) (those are 64 bits only).

It's pretty basic at this time, only used to prevent flickering.

**Comment by yagomont** [comment]
I guess that not redrawing the taskbar all the time also means savings in processing power? This might be big. Not waking up the cpu/gpu to redraw as often might reduce overall power draw.

**Comment by denosawr** [comment]
@yagomont Indeed, this should make CPU consumption and battery drain negligible :)

**Comment by sylveon** [comment]
(it already is pretty negligible)

**Comment by sylveon** [comment]
Dev builds hook into explorer.
